realm/realm-library/src/androidTest/kotlin/io/realm/RealmModelManagedSetTester.kt

```diff
@@ -0,0 +1,397 @@
+            notPresentValueSet.add(unmanagedNotPresentValue)
+            assertTrue(set.addAll(notPresentValueSet as Collection<T>))
+            assertEquals(unmanagedInitializedSet.size + notPresentValueSet.size, set.size)
+
+            // Does not change after adding a managed list with the same elements
+            set.clear()
+            set.addAll(unmanagedInitializedSet)
+            val sameValuesManagedList = managedCollectionGetter.call(transactionRealm.createObject<SetContainerClass>())
+            sameValuesManagedList.addAll(unmanagedInitializedSet)
+            assertFalse(set.addAll(sameValuesManagedList))
+            assertTrue(set.containsAll(sameValuesManagedList))
+
+            // Changes after adding a managed list with other elements
+            val differentValuesManagedList = managedCollectionGetter.call(transactionRealm.createObject<SetContainerClass>())
+            differentValuesManagedList.addAll(listOf(unmanagedNotPresentValue))
+            assertTrue(set.addAll(differentValuesManagedList))
+            assertTrue(set.containsAll(differentValuesManagedList))
+
+            // Does not change after adding an empty managed list
+            set.clear()
+            assertTrue(set.addAll(unmanagedInitializedSet))
+            val emptyValuesManagedList = managedCollectionGetter.call(transactionRealm.createObject<SetContainerClass>())
+            assertFalse(set.addAll(emptyValuesManagedList))
+            assertEquals(unmanagedInitializedSet.size, set.size)
+
+            // Fails if passed null according to Java Set interface
+            assertFailsWith<NullPointerException> {
+                set.addAll(TestHelper.getNull())
+            }
+        }
+
+        // Test with objects from another realm
+        looperThread.runBlocking {
+            Realm.getInstance(realm.configuration).use { alternativeRealm ->
+                alternativeRealm.executeTransaction {
+                    val values = alternativeRealm.copyToRealmOrUpdate(unmanagedInitializedSet)
+
+                    assertFailsWith<IllegalArgumentException>("Cannot pass an object from another Realm instance") {
+                        set.addAll(values)
+                    }
+                }
+            }
+            looperThread.testComplete()
+        }
+    }
+
+    override fun retainAll() {
+        // Test with managed realm objects
+        managedTester.retainAll()
+
+        // Test with unmanaged realm objects
+        val set = initAndAssertEmptySet()
+        realm.executeTransaction {
+            // Check throws exception when unmanaged values are passed
+            assertFailsWith<IllegalArgumentException>("Collection with unmanaged objects not permitted") {
+                set.retainAll(unmanagedInitializedSet)
+            }
+
+            // Check throws exception when null values are passed
+            assertFailsWith<IllegalArgumentException>("Collections with nulls are not permitted") {
```

Comment:
Following Eduardo's comment, I guess we should throw a NullPointerException to follow Java's Set API.